cursorless.nvim/test/unit/cursorless_spec.lua

```diff
@@ -0,0 +1,95 @@
+describe("cursorless tests", function()
+  local cursorless = require("cursorless.cursorless")
+
+  local get_selected_text = function()
+    local _, ls, cs = unpack(vim.fn.getpos("v"))
+    local _, le, ce = unpack(vim.fn.getpos("."))
+    return vim.api.nvim_buf_get_text(0, ls - 1, cs - 1, le - 1, ce, {})
+  end
+
+  local function convert_table_entries(tbl, func)
+    local mapped = {}
+    for k, v in pairs(tbl) do
+      mapped[k] = func(v)
+    end
+    return mapped
+  end
+
+  describe("window_get_visible_lines()", function()
+    before_each(function() end)
+
+    it("Can read one visible line", function()
+      local pos = vim.api.nvim_win_get_cursor(0)[2]
+      local line = vim.api.nvim_get_current_line()
+      local nline = line:sub(0, pos) .. "hello" .. line:sub(pos + 1)
+      vim.api.nvim_set_current_line(nline)
+
+      local visible = cursorless.window_get_visible_lines()
+      assert(table.concat(visible) == table.concat({ 1, 1 }))
+    end)
+
+    it("Can read all lines visible on the window", function()
+      local maxlines = vim.api.nvim_win_get_height(0)
+      local lines = {}
+      for _ = 1, (maxlines + 1) do
+        table.insert(lines, "hello ")
+      end
+      vim.api.nvim_buf_set_lines(0, 0, -1, false, lines)
+      local visible = cursorless.window_get_visible_lines()
+      assert(table.concat(visible) == table.concat({ 1, maxlines }))
+    end)
+  end)
+  describe("select_range()", function()
+    it("Selects the specified range", function()
+      local lines = "hello world"
+      vim.api.nvim_buf_set_lines(0, 0, -1, false, vim.split(lines, "\n"))
+      cursorless.select_range(1, 2, 1, 4)
+
+      assert(table.concat(get_selected_text()) == "llo")
+    end)
+  end)
+  describe("buffer_get_selection()", function()
+    it(
+      "Can get the forward selection in a format expected by cursorless",
+      function()
+        local lines = "hello world"
+        vim.api.nvim_buf_set_lines(0, 0, -1, false, vim.split(lines, "\n"))
+        cursorless.select_range(1, 2, 1, 4)
+        assert(
+          table.concat(
+            convert_table_entries(cursorless.buffer_get_selection(), tostring),
+            ", "
+          )
+            == table.concat(
+              convert_table_entries({ 1, 3, 1, 5, false }, tostring),
+              ", "
+            )
+        )
+      end
+    )
+    it(
+      "Can get the backward selection in a format expected by cursorless",
+      function()
+        local lines = "hello world"
+        vim.api.nvim_buf_set_lines(0, 0, -1, false, vim.split(lines, "\n"))
+        cursorless.select_range(1, 4, 1, 2)
```

Comment:
I guess we should add a test for each helper function and have those tests run first, so it's more apparent that a subsequent complex test that relies on the helper is possibly failing because of the earlier failure? Didn't address that yet

Reply:
Happy if you want to create an issue for that and do it in a future PR or if you want to do it as part of this PR.

Reply:
I'll likely just do a follow up ya

Reply:
Filed an issue here: https://github.com/cursorless-dev/cursorless/issues/2548